torchvision/utils.py

```diff
@@ -169,6 +169,8 @@ def draw_bounding_boxes(
         colors (color or list of colors, optional): List containing the colors
             of the boxes or single color for all boxes. The color can be represented as
             PIL strings e.g. "red" or "#FF00FF", or as RGB tuples e.g. ``(240, 10, 157)``.
+            By default, random colors are generated for boxes.
+            If labels are provided, boxes with same labels have same color.
```

Comment:
I'm concerned with this API change because the behaviour changes depending on whether the labels are defined. Effectively the new API tries to do things for the user instead of letting them decide what to do. I think it would be best to stick to generating a random palette for as many masks as we detected. This will be similar to what we do on the segmentation drawing tool and will enable us to reuse methods such as `_generate_color_palette()`.

Reply:
@datumbox would you mind detailing what you mean by "enable us to reuse methods such as _generate_color_palette()" ?

Having one box color per label is a default behaviour that makes sense to me, and @oke-aditya followed my earlier suggestion of https://github.com/pytorch/vision/issues/4528#issuecomment-933232255. It avoids users some extra code.

`draw_segmentation_masks()` doesn't accept a `label` parameter, so it's not unexpected that its behaviour differs from `draw_bounding_boxes()` in that respect.

Reply:
@NicolasHug I was thinking that `utils._generate_color_palette(num_masks)` can be used to generate a colour per mask (not per label) as currently happens on the segmentation case. 

Concerning the lack of support for labels on segmentation, this is indeed the situation with our utility but doesn't have to be the case going forwards. Other libraries such as [detectron2](https://github.com/facebookresearch/detectron2) offer visualizations for segmentation and masks that display the labels.

TBH I haven't followed the original discussions (I try to catch up coming back from PTO). My thoughts on this is that if we want to keep the utility simple, we could let the users define one colour per mask. If they need to apply a more complex logic (such as keeping the same colour per label), this should be done outside of the utility using their own custom logic and provide the list of colours to our util. This would align the approach across segmentation and detection. Let me know your thoughts.

Reply:
> Concerning the lack of support for labels on segmentation, this is indeed the situation with our utility but doesn't have to be the case going forwards. Other libraries such as detectron2 offer visualizations for segmentation and masks that display the labels.

I agree. And if we were to support labels in `draw_segmentation_masks()`, I think it would also make sense to have one color per label by default :)

I also agree that complexity is a concern, both regarding the code and regarding what the user would expect from that function. As far as I can tell the only extra code that handling labels adds are these few lines below:


```py
        if labels is not None:
            label_color_map = dict(zip(labels, colors))
            colors = [label_color_map[label] for label in labels]
```

which I feel is minimal.

Reply:
Users still can draw colors for their labels. If both labels and colors are passed. The `ith` box would get the `ith` color, irrespective of the labels.
I guess the current improvement just improves how the the default colors are handled. The need to pass a color palette reduces.